core/test/solver/batch_cg.cpp

```diff
@@ -0,0 +1,277 @@
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::share(gko::test::generate_3pt_stencil_batch_matrix<Mtx>(
+              this->exec->get_master(), num_batch_items, num_rows))),
+          solver_factory(Solver::build()
+                             .with_max_iterations(def_max_iters)
+                             .with_tolerance(def_abs_res_tol)
+                             .with_tolerance_type(def_tol_type)
+                             .on(exec)),
+          solver(solver_factory->generate(mtx))
+    {}
+
+    std::shared_ptr<const gko::Executor> exec;
+    const gko::size_type num_batch_items = 3;
+    const int num_rows = 5;
+    std::shared_ptr<const Mtx> mtx;
+    const int def_max_iters = 100;
+    const real_type def_abs_res_tol = 1e-11;
+    const gko::batch::stop::tolerance_type def_tol_type =
+        gko::batch::stop::tolerance_type::absolute;
+    std::unique_ptr<typename Solver::Factory> solver_factory;
+    std::unique_ptr<gko::batch::BatchLinOp> solver;
+};
+
+TYPED_TEST_SUITE(BatchCg, gko::test::ValueTypes, TypenameNameGenerator);
+
+
+TYPED_TEST(BatchCg, FactoryKnowsItsExecutor)
+{
+    ASSERT_EQ(this->solver_factory->get_executor(), this->exec);
+}
+
+
+TYPED_TEST(BatchCg, FactoryHasCorrectDefaults)
+{
+    using Solver = typename TestFixture::Solver;
+    using Mtx = typename TestFixture::Mtx;
+    using value_type = typename TestFixture::value_type;
+
+    auto solver_factory = Solver::build().on(this->exec);
+    auto solver = solver_factory->generate(Mtx::create(this->exec));
+
+    ASSERT_NE(solver->get_system_matrix(), nullptr);
+    ASSERT_NE(solver->get_preconditioner(), nullptr);
+    ASSERT_NO_THROW(gko::as<gko::batch::matrix::Identity<value_type>>(
+        solver->get_preconditioner()));
+    ASSERT_EQ(solver->get_tolerance(), 1e-11);
+    ASSERT_EQ(solver->get_max_iterations(), 100);
+    ASSERT_EQ(solver->get_tolerance_type(),
+              gko::batch::stop::tolerance_type::absolute);
+}
+
+
+TYPED_TEST(BatchCg, FactoryCreatesCorrectSolver)
+{
+    using Solver = typename TestFixture::Solver;
+    ASSERT_EQ(this->solver->get_common_size(),
+              gko::dim<2>(this->num_rows, this->num_rows));
+
+    auto solver = gko::as<Solver>(this->solver.get());
+
+    ASSERT_NE(solver->get_system_matrix(), nullptr);
+    ASSERT_EQ(solver->get_system_matrix(), this->mtx);
+}
+
+
+TYPED_TEST(BatchCg, CanBeCopied)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Solver = typename TestFixture::Solver;
+    auto copy = this->solver_factory->generate(Mtx::create(this->exec));
+
+    copy->copy_from(this->solver.get());
+
+    ASSERT_EQ(copy->get_common_size(),
+              gko::dim<2>(this->num_rows, this->num_rows));
+    ASSERT_EQ(copy->get_num_batch_items(), this->num_batch_items);
+    auto copy_mtx = gko::as<Solver>(copy.get())->get_system_matrix();
+    const auto copy_batch_mtx = gko::as<const Mtx>(copy_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), copy_batch_mtx, 0.0);
```

Comment:
I guess one could also test that the pointers are not the same.

Reply:
The pointers will be the same as the only a shallow copy is done. It is the same case in the non-batched as well.

Reply:
I guess that's true. Maybe something to think about in the future.